### PR TITLE
[Api][Checkout] ShippingMethod on Checkout

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        return $this->getCheckoutShippingMethods($cart, $shipment);
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return
+            is_a($resourceClass, ShippingMethodInterface::class, true) &&
+            !isset($subresourceIdentifiers['id']) &&
+            !isset($subresourceIdentifiers['shipments'])
+        ;
```
Comment (Tomanhez) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php:72]: This condition allows using this SubresourceDataProvider only for `/new-api/orders/{id}/shipments/{shipments}/methods` endpoint

File: src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShipmentInterface;
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Core\Repository\ShipmentRepositoryInterface;
+use Sylius\Component\Registry\ServiceRegistryInterface;
+use Sylius\Component\Shipping\Resolver\CompositeMethodsResolver;
+use Webmozart\Assert\Assert;
+
+final class ShippingMethodsSubresourceDataProvider implements RestrictedDataProviderInterface, SubresourceDataProviderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php:27]:
```suggestion
final class CartShippingMethodsSubresourceDataProvider implements RestrictedDataProviderInterface, SubresourceDataProviderInterface
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        $shippingMethods = $this->compositeMethodsResolver->getSupportedMethods($shipment);
+
+        /** @var ShippingMethodInterface $shippingMethod */
+        foreach ($shippingMethods as $shippingMethod)
+        {
+            $calculator = $this->calculators->get($shippingMethod->getCalculator());
+            /** @var int $amount */
+            $amount = $calculator->calculate($shipment, $shippingMethod->getConfiguration());
+
+            $resolvedShippingMethods[] = ['shippingMethod' => $shippingMethod, 'calculatedAmount' => $amount];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ShippingMethodsSubresourceDataProvider.php:92]:
```suggestion
            $resolvedShippingMethods[] = ['shippingMethod' => $shippingMethod, 'cost' => $amount];
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        $subresourceIdentifiers = isset($context['subresource_identifiers']) ? $context['subresource_identifiers'] : null;
+
+        return
+            is_a($resourceClass, ShippingMethodInterface::class, true) &&
+            isset($subresourceIdentifiers['id']) &&
+            isset($subresourceIdentifiers['shipments'])
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php:83]: i think we can merge it into one `isset($subresourceIdentifiers['id'], $subresourceIdentifiers['shipments'])`

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -293,6 +330,14 @@ public function myOrdersShippingMethodShouldBe(ShippingMethodInterface $shipping
         );
     }
 
+    /**
+     * @Then I should be informed that my order cannot be shipped to this address
+     */
+    public function iShouldBeInformedThatMyOrderCannotBeShippedToThisAddress(): void
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:336]: Maybe merge it with line 368? :D they do the same ( just suggesting)

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    public function getSubresource(string $resourceClass, array $identifiers, array $context, string $operationName = null)
+    {
+        $subresourceIdentifiers = $context['subresource_identifiers'];
+
+        /** @var OrderInterface|null $cart */
+        $cart = $this->orderRepository->findCartByTokenValue($subresourceIdentifiers['id']);
+        Assert::notNull($cart);
+
+        /** @var ShipmentInterface $shipment */
+        $shipment = $this->shipmentRepository->findOneBy(['id' => $subresourceIdentifiers['shipments']]);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php:68]:
```suggestion
        $shipment = $this->shipmentRepository->find($subresourceIdentifiers['shipments']);
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        Assert::notNull($shipment);
+
+        Assert::true($cart->hasShipment($shipment), 'Shipment doesn\'t match for order');
+
+        return $this->getCartShippingMethods($cart, $shipment);
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        $subresourceIdentifiers = isset($context['subresource_identifiers']) ? $context['subresource_identifiers'] : null;
```
Comment (arti0090) [inline src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php:78]: could be used `$subresourceIdentifiers = $context['subresource_identifiers'] ?? null;`

File: src/Sylius/Bundle/ApiBundle/DTO/CartShippingMethod.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\DTO;
+
+use Sylius\Component\Core\Model\ShippingMethodInterface;
+
+class CartShippingMethod implements CartShippingMethodInterface
+{
+    /** @var string */
+    private $code;
```
[the remark this answers is not in the document]
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/DTO/CartShippingMethod.php:21]: We need ID for api_resource configuration https://github.com/Sylius/Sylius/pull/11735/files#diff-245e91664a059e65a6c342f51b7bd377R8

File: src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php (hunk 7 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/DataProvider/CartShippingMethodsSubresourceDataProvider.php:78]: no, because we need use isset() for check existing key

**Comment by lchrusciel** [comment]
Thanks, Tomasz! :1st_place_medal:

**Pull request merged**